cmd/kubeadm/app/util/config/cluster_test.go

```diff
@@ -727,41 +825,275 @@ func TestGetGetClusterStatus(t *testing.T) {
 			client := clientsetfake.NewSimpleClientset()
 
 			for _, c := range rt.configMaps {
-				err := c.create(client)
+				err := c.Create(client)
 				if err != nil {
-					t.Errorf("couldn't create ConfigMap %s", c.name)
+					t.Errorf("couldn't create ConfigMap %s", c.Name)
 					return
 				}
 			}
 
-			clusterStatus, err := GetClusterStatus(client)
+			_, err := GetClusterStatus(client)
 			if rt.expectedError != (err != nil) {
 				t.Errorf("unexpected return err from GetClusterStatus: %v", err)
 				return
 			}
 			if rt.expectedError {
 				return
 			}
-
-			// Test expected values in clusterStatus
-			if len(clusterStatus.APIEndpoints) != rt.expectedEndpoints {
-				t.Errorf("unexpected ClusterStatus return value")
-			}
 		})
 	}
 }
 
-type fakeConfigMap struct {
-	name string
-	data map[string]string
+func TestGetAPIEndpointFromPodAnnotation(t *testing.T) {
+	var tests = []struct {
+		name             string
+		nodeName         string
+		backoff          wait.Backoff
+		pods             []testresources.FakeStaticPod
+		clientSetup      func(*clientsetfake.Clientset)
+		expectedEndpoint kubeadmapi.APIEndpoint
+		expectedErr      bool
+	}{
+		{
+			name:     "exactly one pod with annotation",
+			nodeName: nodeName,
+			backoff:  wait.Backoff{Duration: 0, Jitter: 0, Steps: 1},
```

Comment:
nit: I don't think test testing the backoff logic should be done here, this is a utility already tested elsewhere

Reply:
We are note testing the backoff logic. We are testing [this logic](https://github.com/kubernetes/kubernetes/pull/87656/files#diff-60f768874354bccc5862ca90ec601fc6R210-R228), controlling the backoff, so the wait on retries is faster while running unit tests.

Reply:
it's really odd to have unit tests with a zero backoff.

ideally, there should be a function without backoff that is unit tested. the function that wraps it should have a backoff (named `...WithRetry`) and should not be unit tested.

Reply:
Let me elaborate a little more: the idea behind this was to have fast unit test executions even when playing with retries. If we are testing how our function fallbacks, we want to mimic certain retry logic, and I'd like tests to not sleep for seconds.

Ideally, the only place where we want to control backoff on unit tests is on the functions `getAPIEndpointWithBackoff` and `getEtcdEndpointsWithBackoff`, so we cover the case in which the information is not present in an annotation (or apiserver temporarily unreachable), and so we try to fallback to the cluster status.

Does this make sense? If it does, I will update the tests, and simplify tests where backoff is not providing a real benefit (e.g. `TestGetAPIEndpointFromPodAnnotation` can get rid of the `backoff` in the test table).

Reply:
i can understand the idea to have better coverage.

do we really care to pass a custom backoff from each sub-test, or can it be a variable defined near the getAPIEndpointFromPodAnnotation() call?

Reply:
It can totally be on the code block that calls the function, instead of part of the test table. Maybe I have to keep it as part of the test table (each subtest) if I'm going to test the fallback (so the waits are reasonable when unit testing).

I'll update this PR.